lib/collection/src/shards/local_shard.rs

```diff
@@ -398,21 +398,36 @@ impl LocalShard {
         let segments = self.segments();
         // ToDo: Start from minimal applied version
         for (op_num, update) in wal.read_all() {
-            // Panic only in case of internal error. If wrong formatting - skip
-            if let Err(CollectionError::ServiceError { error, backtrace }) =
-                CollectionUpdater::update(segments, op_num, update)
-            {
-                if let Some(backtrace) = backtrace {
-                    log::error!("Backtrace: {}", backtrace);
+            // Propagate `CollectionError::ServiceError`, but skip other error types.
+            match CollectionUpdater::update(segments, op_num, update) {
+                Err(CollectionError::ServiceError { error, backtrace }) => {
+                    let path = self.path.display();
+
+                    log::error!(
+                        "Can't apply WAL operation: {error}, \
+                         collection: {collection_id}, \
+                         shard: {path}, \
+                         op_num: {op_num}"
+                    );
+
+                    if let Some(backtrace) = &backtrace {
+                        log::error!("Backtrace: {}", backtrace);
+                    }
+
+                    // TODO: This is ugly AF. "But sTRuCt VaRiaNts ArE pReTTy aND BetTeR CarRy SeMaNtICs!" :/
+                    return Err(CollectionError::ServiceError { error, backtrace });
                 }
-                let path = self.path.display();
-                panic!("Can't apply WAL operation: {error}, collection: {collection_id}, shard: {path}, op_num: {op_num}");
+
+                Err(err) => log::error!("{err}"),
+                Ok(_) => (),
             }
             bar.inc(1);
         }
 
-        self.segments.read().flush_all(true).unwrap();
+        self.segments.read().flush_all(true)?;
```

Comment:
Does this result in handling the error properly? In other words: does this now help prevent the crash-on-start loop?

I didn't dive into the details here.

Reply:
Yep. That's a part of it.